imageio/plugins/pyav.py

```diff
@@ -0,0 +1,839 @@
+
+PyAVPlugin is a plugin that wraps pyAV. pyAV is a set of pythonic bindings for
+the FFMPEG library. As such this plugin is similar our famous FFMPEG plugin, but
+offers nicer bindings and aims to superseed it in the future.
+
+"""
+
+from math import ceil
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import av
+import av.filter
+import numpy as np
+from numpy.typing import ArrayLike
+from numpy.lib.stride_tricks import as_strided
+
+from ..core import Request
+from ..core.request import InitializationError, IOMode, URI_BYTES
+from ..core.v3_plugin_api import PluginV3, ImageProperties
+
+
+def _format_to_dtype(format: av.VideoFormat) -> np.dtype:
+    """Convert a pyAV video format into a numpy dtype"""
+
+    if len(format.components) == 0:
+        # fake format
+        raise ValueError(
+            f"Can't determine dtype from format `{format.name}`. It has no channels."
+        )
+
+    endian = ">" if format.is_big_endian else "<"
+    dtype = "f" if "f32" in format.name else "u"
+    bits_per_channel = [x.bits for x in format.components]
+    n_bytes = str(int(ceil(bits_per_channel[0] / 8)))
+
+    return np.dtype(endian + dtype + n_bytes)
+
+
+def _get_frame_shape(frame: av.VideoFrame) -> Tuple[int, ...]:
+    """Compute the frame's array shape
+
+    Parameters
+    ----------
+    frame : av.VideoFrame
+        A frame for which the resulting shape should be computed.
+
+    Returns
+    -------
+    shape : Tuple[int, ...]
+        A tuple describing the shape of the image data in the frame.
+
+    Notes
+    -----
+
+    This function assumes that channel components will be byte-aligned upstream
+    if necessary. This applies to the formats monow, monob, rgb4, and rgb8.
+    (This will be converted to gray8/rgb24 upstream.)
+
+    """
+
+    widths = [component.width for component in frame.format.components]
+    heights = [component.height for component in frame.format.components]
+    bits = np.array([component.bits for component in frame.format.components])
+    line_sizes = [plane.line_size for plane in frame.planes]
+
+    subsampled_width = widths[:-1] != widths[1:]
+    subsampled_height = heights[:-1] != heights[1:]
+    unaligned_components = np.any(bits % 8 != 0) or (line_sizes[:-1] != line_sizes[1:])
+    if subsampled_width or subsampled_height or unaligned_components:
+        raise IOError(
+            f"{frame.format.name} can't be expressed as a strided array."
+            "Use `format=` to select a format to convert into."
+        )
+
+    shape = [frame.height, frame.width]
+
+    n_planes = max([component.plane for component in frame.format.components]) + 1
+    if n_planes > 1:
+        shape = [n_planes] + shape
+
+    channels_per_plane = [0] * n_planes
+    for component in frame.format.components:
+        channels_per_plane[component.plane] += 1
+    n_channels = max(channels_per_plane)
+
+    if n_channels > 1:
+        shape = shape + [n_channels]
```

Comment:
I'm a bit confused. This suggests that its always `height, width, channels`, but one of the tests seems to suggest it does not. I'm probably missing something.

Reply:
Hm, maybe the name `channel` is a bit misleading, but I'm not 100% sure what good taxonomy for this should be. 

FFMpeg stores frames as a sequence of image planes and pixel components (`channels`) can be spread out between these planes. Some formats use one plane and a pixel's components are packed/contigous (aka, the format is channel-last). For these, `n_channels` counts how many color channels the format has. Other formats store one component per plane (aka the format is channel-first) and for those formats `n_planes` counts the number of channels the format has.

Finally, there are esoteric formats. `n12` and consorts store a different number of of components per plane; this is not trivially strideable so we can't easily create a ndarray from this and I'm raising for this format. There are also `bayer_rggb8` and consorts, which are stridable, but pixels aren't packed so the notion of `channels` inside `shape` doesn't make sense anymore. I'm currently raising for this format as well, but I'm very open for suggestions on how to support it (because it is possible, but the question is how to do so cleanly).

Reply:
Maybe just put what you just said in a comment :)

Reply:
Done.